src/Features/Core/Portable/ConvertAnonymousTypeToClass/AbstractConvertAnonymousTypeToClassCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,412 @@
+                }
+
+                if (anonymousType.Equals(childType))
+                {
+                    ReplaceWithObjectCreation(editor, className, creationNode, childCreation);
+                }
+            }
+        }
+
+        private void ReplaceWithObjectCreation(
+            SyntaxEditor editor, string className,
+            TAnonymousObjectCreationExpressionSyntax startingCreationNode,
+            TAnonymousObjectCreationExpressionSyntax childCreation)
+        {
+            // Use the callback form as anonymous types may be nested, and we want to
+            // properly replace them even in that case.
+            editor.ReplaceNode(
+                childCreation,
+                (currentNode, g) =>
+                {
+                    var currentAnonymousObject = (TAnonymousObjectCreationExpressionSyntax)currentNode;
+
+                    // If we hit the node the user started on, then add the rename annotation here.
+                    var classNameToken = startingCreationNode == childCreation
+                        ? g.Identifier(className).WithAdditionalAnnotations(RenameAnnotation.Create())
+                        : g.Identifier(className);
+
+                    var classNameNode = (TIdentifierNameSyntax)g.IdentifierName(classNameToken);
+                    return CreateObjectCreationExpression(classNameNode, currentAnonymousObject)
+                        .WithAdditionalAnnotations(Formatter.Annotation);
+                });
+        }
+
+        private async Task<INamedTypeSymbol> GenerateFinalNamedTypeAsync(
+            Document document, string className,
+            ImmutableArray<IPropertySymbol> properties, CancellationToken cancellationToken)
+        {
+            var compilation = await document.Project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+
+            // Fext, see if any of the properties ended up using any type parameters from the
+            // containing method/named-type.  If so, we'll need to generate a generic type so we can
+            // properly pass these along.
+            var capturedTypeParameters =
+                properties.Select(p => p.Type)
+                          .SelectMany(t => t.GetReferencedTypeParameters())
+                          .Distinct()
+                          .ToImmutableArray();
+
+            // Now try to generate all the members that will go in the new class. This is a bit
+            // circular.  In order to generate some of the members, we need to know about the type.
+            // But in order to create the type, we need the members.  To address this we do two
+            // passes. First, we create an empty version of the class.  This can then be used to
+            // help create members like Equals/GetHashCode.  Then, once we have all the members we
+            // create the final type.
+            var namedTypeWithoutMembers = CreateNamedType(className, capturedTypeParameters, default);
```

Comment:
>default [](start = 93, length = 7)

Should we pass the cancellation token in?

Reply:
Nope.  there's nothing to cancel.  and it's basically jsut allocating a simple object.